### PR TITLE
Asset annotation

### PR DESCRIPTION
**Pull request by robes**
Specifies the 2016-Asset annotation.

### CODE REVIEW

File: user-doc/annotation.md [hunk 1]
```diff
@@ -401,12 +402,53 @@ the alternative one has modified the representation of each entity in some way.
...
+
+`tag:isrd.isi.edu,2016:asset`
+
+This key indicates that the annotated table may be interpreted as asset metadata.
+
+Supported JSON payload patterns:
+
+- `null` or `{}`: Default heuristics apply.
+- `{`... `"internal":` [_column_, ...] ...`}`: The one or more named _columns_ store internal identifiers for the asset, used for efficient normalized storage in the database but not meaningful to typical users. The referenced columns MUST each comprise a single-column key for the table.
+- `{`... `"identifier": ` _column_ ...`}`: The _column_ stores a resolvable, globally unique, identifier for the asset. In its simplest form this MAY be satisfied by supplying a URL to the resource. The identifier MUST be _globally unique_ which is satisfiable by any fully qualified HTTP URL, for example. In any case the identifier MUST be _resolvable_ meaning that a client can retrieve the identified asset either directly or indirectly through a resolution service. More formal identifier schemes may also be considered such as MINID, DOI, ARK, or PURL. It is up to client tooling to recognize and resolve identifiers in such schemes.
```
Comment (karlcz) [inline user-doc/annotation.md]: I think we can say the column values *denote* a globally unique identifier but cannot require that they *contain* such an identifier.  E.g. a relative URI should be allowed as the missing scheme and authority part would come from applying a normal web standards interpretation to the ermrest API response.

File: user-doc/annotation.md [hunk 2]
```diff
@@ -401,12 +402,53 @@ the alternative one has modified the representation of each entity in some way.
...
+- `{`... `"accession_number": ` _column_ ...`}`: The _column_ stores the "accession number" for the asset. The scheme for an accession number is defined by an application (i.e., a science data repository, data hub, community data resource service, etc.). The accession number MUST be guaranteed to be unique _within_ the application. They MAY be globally unique but generally are not, in practice. As an example, accession numbers often take the following form `PREFIX + SERIAL NUMBER [ '.' + REVISION NUMBER ]`, such as `GDS6248`. The example given here is non-normative.
+- `{`... `"name": ` _pattern_ ...`}`: A name for the asset to be derived by a [Pattern Expansion](#pattern-expansion) on _pattern_. The name may be used to represent a file name, but is not required to do so.
+- `{`... `"bytes": ` _column_ ...`}`: The _column_ stores the file size in bytes of the asset. It should be an integer typed column.
+- `{`... `"sha256": ` _column_ ...`}`: The _column_ stores the checksum generated by the 'sha256' cryptographic hash function.
```
Comment (karlcz) [inline user-doc/annotation.md]: This needs to be more precise about encoding. The algorithm produces a bit vector which is commonly encoded as a hexadecimal digit string using ASCII/UTF-8 character set.  But some other standards do things like a base64 encoding of the bit vector interpreted as an octet string.

File: user-doc/annotation.md [hunk 3]
```diff
@@ -401,12 +402,53 @@ the alternative one has modified the representation of each entity in some way.
...
+- `{`... `"creator": ` _column_ ...`}`: The _column_ stores the _creator_ of the asset. The _column_ MAY be defined as a foreign key that references a person or user table in the model. The usual model introspection applies.
+- `{`... `"description": ` _pattern_ ...`}`: A description for the asset to be derived by a [Pattern Expansion](#pattern-expansion) on _pattern_.
+- `{`... `"date": ` _column_ ...`}`: The _column_ stores the date that the asset metadata was recorded in the system.
+- `{`... `"mediatype": ` _column_ ...`}`: The _column_ stores the Media Type of the asset, for example `application/dicom`. The format of its contents MUST conform to the [Media Type RFC](https://tools.ietf.org/html/rfc6838). The values used SHOULD come from the [list of Media Types](http://www.iana.org/assignments/media-types/media-types.xhtml) registered with IANA.
+- `{`... `"format": ` _column_ ...`}`: The _column_ stores the identifier for the file format that the asset _conforms to_, for example `http://edamontology.org/format_3548`. See the [EDAM Ontology](http://edamontology.org) for a listing of file formats used in bioinformatics.
```
Comment (karlcz) [inline user-doc/annotation.md]: Shouldn't we say these must be globally unique URNs or similar?

File: user-doc/annotation.md [hunk 4]
```diff
@@ -401,12 +402,53 @@ the alternative one has modified the representation of each entity in some way.
...
+- `{`... `"contributor": ` _column_ ...`}`: The _column_ stores the "contributors" of the asset. The concept of contributor is not strictly defined here and left up to the application to define the exact criteria for what constitutes a contributor. When this key is present in the JSON payload the _column_ should be a JSON list of contributors. If this key is not present, applications should look for an inbound foreign key reference with a `from_name` annotation set to `'contributor'`, or the application should look for a relationship via a binary association table with a `to_name` annotation set to `'contributor'`. The entities of that related table should be interpreted as the set of contributors to the asset.
```
Comment (karlcz) [inline user-doc/annotation.md]: This should also allow an explicit `[schema, constraint]` name-pair for the specific foreign key that links contributors.

File: user-doc/annotation.md [hunk 5]
```diff
@@ -401,12 +402,53 @@ the alternative one has modified the representation of each entity in some way.
...
+
+#### Required Properties
+
+The following table specifies the required properties for the annotation. We also recommend additional properties for enhanced functionality provided by some client tooling.
+
+| Property     | Required | Recommended |
+|:-------------|:--------:|:-----------:|
+| `identifier` | X | - |
+| `mediatype`  | X | - |
```
Comment (karlcz) [inline user-doc/annotation.md]: why is media type required?

File: user-doc/annotation.md [hunk 6]
```diff
@@ -401,12 +402,53 @@ the alternative one has modified the representation of each entity in some way.
...
+| `bytes`      | - | X |
+| `sha256`     | - | X |
+| `format`     | - | X |
+| `name`       | - | X |
+
+#### Heuristics
+
+1. In the absence of this annotation, assume that a table represents assets if it defines the required set of columns for this annotation (i.e., `identifier` and `mediatype`).
```
Comment (karlcz) [inline user-doc/annotation.md]: Via case-insensitive match?

File: user-doc/annotation.md (hunk 1 above)
Comment (robes) [inline user-doc/annotation.md]: Right, we discussed that yesterday. @mikedarcy does that work for your client tooling? I assume we can make it work even if not.

File: user-doc/annotation.md (hunk 2 above)
Comment (robes) [inline user-doc/annotation.md]: we should just pick and specify the most common and easy to produce encoding. @mikedarcy in fact, what is the encoding needed by Bags?

File: user-doc/annotation.md (hunk 3 above)
Comment (robes) [inline user-doc/annotation.md]: we could say 'globally unique identifier'. i don't know whether it should be specified as a URN or not.

File: user-doc/annotation.md (hunk 4 above)
Comment (robes) [inline user-doc/annotation.md]: right

File: user-doc/annotation.md (hunk 5 above)
Comment (robes) [inline user-doc/annotation.md]: this should be revisited. i was thinking that we want to heuristically determine that a table represents assets and to do that we need more than an 'identifier'. so there are a few options:
- 'identifier' actually seems like the only required field (if you have the asset annotation on the table)
- we could then say that the annotation itself is _required_ and we do not heuristically determine that a table represents assets
- or we need to pick some subsets of attributes, greater than just 'identifier', to say we can heuristically identify the table as assets.

File: user-doc/annotation.md (hunk 1 above)
Comment (mikedarcy) [inline user-doc/annotation.md]: It should. I'm interested in seeing an example of the proposed response.

File: user-doc/annotation.md (hunk 2 above)
Comment (mikedarcy) [inline user-doc/annotation.md]: Bags use the common ASCII/UTF-8 hexadecimal encoding as output by md5sum, shasum, etc.

File: user-doc/annotation.md (hunk 1 above)
Comment (karlcz) [inline user-doc/annotation.md]: @mikedarcy not sure what you mean.  A current ERMrest response can have relative URLs (up to the DBA and users what they put in tables) and it should be evaluated relative to the ERMrest URL that was used to get the response.

In practice, we have used absolute paths, e.g. 
- Request is `GET /ermrest/catalog/1/entity/Foo` against `https://www.example.com`
- Therefore `https://www.example.com/ermrest/catalog/1/entity/Foo` is the URL to use as context when interpreting the response
- Response has a URL column containing values like `/hatrac/foo/bar.txt`
- The response values, when interpreted in the request context, mean something like `https://www.example.com/hatrac/foo/bar.txt` etc.

File: user-doc/annotation.md (hunk 5 above)
Comment (robes) [inline user-doc/annotation.md]: no comments on this so I'm going to assume for now that we _do not_ want a heuristic to determine that a table represents assets, based on some subset of columns matches with the properties of the annotation.

**Comment by robes** [comment]
@karlcz please see the updates which ought to address the previous round of comments

File: user-doc/annotation.md [hunk 7]
```diff
@@ -418,37 +418,25 @@ Supported JSON payload patterns:
...
-
-| Property     | Required | Recommended |
-|:-------------|:--------:|:-----------:|
-| `identifier` | X | - |
-| `mediatype`  | X | - |
-| `bytes`      | - | X |
-| `sha256`     | - | X |
-| `format`     | - | X |
-| `name`       | - | X |
+- `{`... `"date": ` _column_ ...`}`: The _column_ stores the date that the asset metadata was recorded in the system. The date MUST be in ISO 8601 format.
```
Comment (karlcz) [inline user-doc/annotation.md]: I think we should require the column to be of a date or timestamp type, rather than talking about encoding.

**Pull request merged**